### PR TITLE
New "LunaticComponents" for orchestrator

### LINKED ISSUE #661: New <LunaticComponents>
Create a generic component that can be used by orchestrators instead of mapping over components

### PR DESCRIPTION
**Pull request by Grafikart**
## Problem

We want lunatic to autofocus the first field, it's a behaviour common on multiple orchestrators. 

- Orchestrators uses an array of components to  mount lunatic and there is no common wrapper around field
- Code within orchestrator (the `components.map()` is a bit to complicated in my taste and duplicated between orchestrators

## Solution

Create a new components that is responsible to display all the fields, while retaining the level of customization required for orchestrator.

### Before 

```jsx
{components.map(function (component) {
  const {
	  id,
	  componentType,
	  response,
	  storeName,
	  conditionFilter,
	  ...other
  } = component;
  const Component = lunatic[componentType];
  
  const storeInfo = storeName ? getStoreInfo(storeName) : {};
  return (
	  <div
		  className="lunatic lunatic-component"
		  key={`component-${id}`}
	  >
		  <Component
			  id={id}
			  response={response}
			  {...other}
			  {...rest}
			  {...component}
			  {...storeInfo}
			  // fill error when needed
			  errors={errorActive[pageTag]}
			  filterDescription={filterDescription}
			  disabled={readOnly}
		  />
	  </div>
  );
  })}
```

### After

```jsx
<LunaticComponents
  autoFocusKey={pageTag}
  components={components}
  componentProps={({ storeName }) => ({
	  errors: errorActive[pageTag],
	  filterDescription: filterDescription,
	  disabled: readOnly,
	  ...(storeName ? getStoreInfo(storeName) : {}),
  })}
  wrapper={({ children, id }) => (
	<div className="lunatic lunatic-component">{children}</div>
  )}
/>
```

- **componentProps**, is a function to pass extra props down to every components (errors for instance)
- **autoFocusKey**, is a key that would trigger an autofocus on change (can be pageTag for instance)
- **wrapper**, is a function to add an extra wrapper around each component

The minimal version would be

```jsx
<LunaticComponents components={components} />
``` 

## Going further

I wonder if we could "hide" lunatic components to encourage consumer to use only this component as an entry point of lunatic

Fix #661

### CODE REVIEW

File: src/components/lunatic-components.tsx [hunk 1]
```diff
@@ -0,0 +1,82 @@
+import React, {
+	Fragment,
+	PropsWithChildren,
+	ReactNode,
```
Comment (ddecrulle) [inline src/components/lunatic-components.tsx:4]: it's type and should be imported as it

File: src/components/lunatic-components.tsx [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+	useEffect,
+	useRef,
+} from 'react';
+import * as lunatic from '../index';
+import { useLunatic } from '../index';
+import { ItemOf } from '../type.utils';
```
Comment (ddecrulle) [inline src/components/lunatic-components.tsx]: type again

File: src/components/lunatic-components.tsx [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+
+type ItemProps = FilledComponent;
+
+function LunaticComponent({ componentType, ...props }: ItemProps) {
+	// Component is too dynamic to be typed
+	const Component = lunatic[componentType as keyof typeof lunatic] as any;
+	return (
+		<div className="lunatic lunatic-component">
+			<Component {...props} />
+		</div>
```
Comment (ddecrulle) [inline src/components/lunatic-components.tsx:77]: This is perhaps redundant with the possibility of having a wrapper. I think it's more intuitive to render the component directly, except to add this div by default in the wrapper function.

**Review by ddecrulle — CHANGES_REQUESTED** [review]

**Review by ddecrulle — APPROVED** [review]
Good ! Take care this is major change, even if orchestrators can still work with the "old way" of orchestration, it's important to support the change by documenting the README

**Review by laurentC35 — APPROVED** [review]
Nice !
It fix also :
- https://github.com/InseeFr/Lunatic/issues/396

**Comment by laurentC35** [comment]
Can you add a minimal doc on README.md about the new feature : `autoFocusKey` ?
Thanks

**Review by laurentC35 — CHANGES_REQUESTED** [review]
Can you add a minimal doc on README.md about the new feature : `autoFocusKey` (like description of your PR)
Thanks

**Pull request merged**